File: 057_SubarraySumEqualsK.py
"""
Topic: Prefix Sum / Hashmap
Problem: Subarray Sum Equals K
Description: Count subarrays whose sum equals k using prefix-sum hashmap.
Time: O(n)
Space: O(n)
"""

from typing import List


def subarraySum(nums: List[int], k: int) -> int:
    from collections import defaultdict
    count = defaultdict(int)
    count[0] = 1
    s = 0
    ans = 0
    for x in nums:
        s += x
        ans += count.get(s - k, 0)
        count[s] += 1
    return ans


if __name__ == '__main__':
    print(subarraySum([1,1,1], 2))  # 2
    print(subarraySum([1,2,3], 3))  # 2

File: 057_SubarraySumEqualsK.cpp
// Topic: Hashmap / Prefix Sum
// Subarray Sum Equals K: count subarrays with sum equal to k using prefix sums and hashmap.
// Time Complexity: O(n)
// Space Complexity: O(n)

#include <bits/stdc++.h>
using namespace std;

int subarraySum(vector<int>& nums, int k)
{
    unordered_map<int, int> cnt;
    cnt[0] = 1;
    int sum = 0, res = 0;
    for (int x : nums)
    {
        sum += x;
        if (cnt.count(sum - k))
            res += cnt[sum - k];
        cnt[sum]++;
    }
    return res;
}

int main()
{
    cout << subarraySum((vector<int>{1, 1, 1}), 2) << "\n";  // 2
    cout << subarraySum((vector<int>{1, 2, 3}), 3) << "\n";  // 2
    return 0;
}
